org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/extension/SonarLintExtensionTracker.java

```diff
@@ -0,0 +1,143 @@
+  private final SonarLintEP<ISonarLintProjectsProvider> projectsProviderEp = new SonarLintEP<>("org.sonarlint.eclipse.core.projectsProvider"); //$NON-NLS-1$
+  private final SonarLintEP<ISonarLintFileAdapterParticipant> fileAdapterParticipantEp = new SonarLintEP<>("org.sonarlint.eclipse.core.fileAdapterParticipant"); //$NON-NLS-1$
+  private final SonarLintEP<ISonarLintProjectFilter> projectFilterEp = new SonarLintEP<>("org.sonarlint.eclipse.core.projectFilter"); //$NON-NLS-1$
+  private final SonarLintEP<IFileLanguageProvider> languageEp = new SonarLintEP<>("org.sonarlint.eclipse.core.languageProvider"); //$NON-NLS-1$
+
+  private static class SonarLintEP<G> {
+
+    private final String id;
+    private final Collection<G> instances = new ArrayList<>();
+
+    public SonarLintEP(String id) {
+      this.id = id;
+    }
+  }
+
+  private final Collection<SonarLintEP<?>> allEps = Arrays.asList(configuratorEp, analysisEp, projectsProviderEp, fileAdapterParticipantEp, projectFilterEp, languageEp);
+
+  private ExtensionTracker tracker;
+
+  public void start() {
+    IExtensionRegistry reg = Platform.getExtensionRegistry();
+    tracker = new ExtensionTracker(reg);
+    IExtensionPoint[] epArray = allEps.stream().map(ep -> reg.getExtensionPoint(ep.id)).toArray(IExtensionPoint[]::new);
+    IFilter filter = ExtensionTracker.createExtensionPointFilter(epArray);
+    tracker.registerHandler(this, filter);
+    for (IExtensionPoint ep : epArray) {
+      for (IExtension ext : ep.getExtensions()) {
+        addExtension(tracker, ext);
+      }
+    }
+  }
+
+  public void close() {
+    if (tracker != null) {
+      tracker.close();
+      tracker = null;
+    }
+  }
+
+  @Override
+  public void addExtension(IExtensionTracker tracker, IExtension extension) {
+    IConfigurationElement[] configs = extension.getConfigurationElements();
+    for (final IConfigurationElement element : configs) {
+      try {
+        for (SonarLintEP ep : allEps) {
+          if (ep.id.equals(extension.getExtensionPointUniqueIdentifier())) {
+            Object instance = element.createExecutableExtension(ATTR_CLASS);
+            ep.instances.add(instance);
+            // register association between object and extension with the tracker
+            tracker.registerObject(extension, instance, IExtensionTracker.REF_WEAK);
+            break;
+          }
+        }
+      } catch (CoreException e) {
+        SonarLintLogger.get().error("Unable to load one SonarLint extension", e);
+      }
+    }
+  }
+
+  @Override
+  public void removeExtension(IExtension extension, Object[] objects) {
+    // stop using objects associated with the removed extension
+    for (SonarLintEP ep : allEps) {
+      if (ep.id.equals(extension.getExtensionPointUniqueIdentifier())) {
+        ep.instances.removeAll(Arrays.asList(objects));
+        break;
+      }
+    }
+  }
+
+  public Collection<ProjectConfigurator> getConfigurators() {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this use of "ProjectConfigurator"; it is deprecated. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)